win32ss/user/winsrv/consrv/condrv/text.c

```diff
@@ -932,138 +933,273 @@ ConDrvReadConsoleOutputString(IN PCONSOLE Console,
+    Length = MultiByteToWideChar(Console->OutputCodePage, 0,
+                                 (PCHAR)StringBuffer,
+                                 NumCodesToWrite,
+                                 NULL, 0);
+    tmpString = RtlAllocateHeap(RtlGetProcessHeap(), 0, Length * sizeof(WCHAR));
+    if (tmpString)
     {
-        return STATUS_INVALID_PARAMETER;
+        WriteBuffer = (LPBYTE)tmpString;
+        MultiByteToWideChar(Console->OutputCodePage, 0,
+                            (PCHAR)StringBuffer,
+                            NumCodesToWrite,
+                            (PWCHAR)WriteBuffer, Length);
+    }
+    else
+    {
+        Status = STATUS_NO_MEMORY;
+        goto Cleanup;
     }
 
-    /* Validity checks */
-    ASSERT(Console == Buffer->Header.Console);
-    ASSERT((StringBuffer != NULL) || (StringBuffer == NULL && NumCodesToWrite == 0));
+    X = WriteCoord->X;
+    Y = (WriteCoord->Y + Buffer->VirtualY) % Buffer->ScreenBufferSize.Y;
 
-    //
-    // FIXME: Make overflow checks on WriteCoord !!!!!!
-    //
+    for (i = 0; i < Length; ++i)
+    {
+        Ptr = ConioCoordToPointer(Buffer, X, Y);
 
-    if (NumCodesWritten) *NumCodesWritten = 0;
+        Ptr->Char.UnicodeChar = *(PWCHAR)WriteBuffer;
+        WriteBuffer += RTL_FIELD_SIZE(CODE_ELEMENT, UnicodeChar);
 
-    switch (CodeType)
-    {
-        case CODE_ASCII:
-            CodeSize = RTL_FIELD_SIZE(CODE_ELEMENT, AsciiChar);
-            break;
+        ++X;
+        if (X == Buffer->ScreenBufferSize.X)
+        {
+            X = 0;
+            ++Y;
+            if (Y == Buffer->ScreenBufferSize.Y)
+            {
+                Y = 0;
+            }
+        }
 
-        case CODE_UNICODE:
-            CodeSize = RTL_FIELD_SIZE(CODE_ELEMENT, UnicodeChar);
-            break;
+        /* For Chinese, Japanese and Korean */
+        if (bCJK && Ptr->Char.UnicodeChar >= 0x80 &&
+            mk_wcwidth_cjk(Ptr->Char.UnicodeChar) == 2)
+        {
+            /* the leading byte */
+            Ptr->Attributes = Buffer->ScreenDefaultAttrib;
+            Ptr->Attributes |= COMMON_LVB_LEADING_BYTE;
+            ++i;
 
-        case CODE_ATTRIBUTE:
-            CodeSize = RTL_FIELD_SIZE(CODE_ELEMENT, Attribute);
-            break;
+            /* the trailing byte */
+            Ptr = ConioCoordToPointer(Buffer, X, Y);
+            Ptr->Attributes = Buffer->ScreenDefaultAttrib;
+            Ptr->Attributes |= COMMON_LVB_TRAILING_BYTE;
 
-        default:
-            return STATUS_INVALID_PARAMETER;
+            ++X;
+            if (X == Buffer->ScreenBufferSize.X)
+            {
+                X = 0;
+                ++Y;
+                if (Y == Buffer->ScreenBufferSize.Y)
+                {
+                    Y = 0;
+                }
+            }
+        }
     }
 
-    if (CodeType == CODE_ASCII)
+Cleanup:
+    if (tmpString)
+        RtlFreeHeap(RtlGetProcessHeap(), 0, tmpString);
+    if (NumCodesWritten)
+        *NumCodesWritten = NumCodesToWrite;
```

Comment:
Seems seem misleading in case of error (in particular in the RtlAllocateHeap failure case)

Reply:
OK.